lms/services/jwt.py

```diff
@@ -0,0 +1,59 @@
+import copy
+import datetime
+
+import jwt
+
+from lms.services.exceptions import ExpiredJWTError, InvalidJWTError
+
+DEFAULT_LIFETIME = datetime.timedelta(hours=1)
+
+
+class JWTService:
+    @classmethod
+    def decode_with_secret(cls, jwt_str, secret) -> dict:
+        """
+        Decode the payload decoded from ``jwt_str``.
+
+        :arg jwt_str: the JWT to decode
+        :arg secret: secret used for decoding
+
+        :raise ExpiredJWTError: if decoding fails because the JWT's timestamp has
+            expired
+        :raise InvalidJWTError: if decoding fails for any other reason (for example
+            ``jwt_str`` is invalid, or ``secret`` is wrong)
+
+        :return: ``jwt_str``'s payload, decoded
+        """
+        try:
+            payload = jwt.decode(
+                jwt_str, secret, algorithms=["HS256"], options={"require": ["exp"]}
+            )
+        except jwt.ExpiredSignatureError as err:
+            raise ExpiredJWTError() from err
+        except jwt.InvalidTokenError as err:
+            raise InvalidJWTError() from err
+
+        del payload["exp"]
+        return payload
+
+    @classmethod
+    def encode_with_secret(cls, payload: dict, secret, lifetime=DEFAULT_LIFETIME):
+        """
+        Encode `payload` as a JWT.
+
+        :arg payload: the payload to encode
+        :arg secret: secret used for encoding
+        :arg lifetime: how long the token should be valid for
+
+        :return: the JWT string
+        """
+        payload = copy.deepcopy(payload)
+        payload["exp"] = datetime.datetime.utcnow() + lifetime
+
+        jwt_str = jwt.encode(payload, secret, algorithm="HS256")
+
+        return jwt_str
+
+
+def factory(_context, _request):
+    return JWTService()
```

Comment:
This is a silly factory but I think we should keep the pattern still. If we need to make it more complex we'll just have to change the factory like in any other service.

Reply:
If we don't need a service factory then it's more efficient not to do it. The default way you are supposed to use `pyramid-services` is to just register an instance the once, which is then re-used on all requests. 

Our default is backwards and we've adopted the per request kind without usually thinking about it. I don't _think_ this was a conscious choice, and we don't have to try and justify it. I think the first few were done this way, and then the rest were copied.

We should probably be moving the other way for those services that don't require it, and then only move towards per-request services when we really have to. If all we need is to read some values from the config, it's not required to read it from the request. We can see an example of this in Checkmate:

 * https://github.com/hypothesis/checkmate/blob/main/checkmate/services/__init__.py
 
My gut feel is a lot of the services which we repeatedly re-create, actually only need to be created once at worker startup.